src/v/archival/archiver_service.cc

```diff
@@ -0,0 +1,999 @@
+#include "cluster/partition_manager.h"
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "container/fragmented_vector.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "model/namespace.h"
+#include "ssx/auto_fmt.h"
+#include "ssx/semaphore.h"
+#include "ssx/sformat.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/util/later.hh>
+
+#include <boost/msm/back/state_machine.hpp>
+#include <boost/msm/front/euml/common.hpp>
+#include <boost/msm/front/functor_row.hpp>
+#include <boost/msm/front/state_machine_def.hpp>
+#include <boost/test/tools/old/interface.hpp>
+#include <fmt/ostream.h>
+
+#include <exception>
+#include <optional>
+#include <utility>
+
+namespace archival {
+
+// This file contains implementation of the managed_partition_stm.
+// The 'managed_partition' is created for every partition running on
+// the shard. The 'managed_partition' then reacts to different events
+// like leadership transfers and partition movement.
+// The events are triggered by the 'partition_manager' and 'group_manager'. Both
+// these services are invoking a callback when the corresponding event happens.
+// The problem is that the callback is synchronous (does not return
+// ss::future<>) but the reaction to the event includes asynchronous calls. E.g.
+// we need to call the 'ntp_archiver::start' and 'ntp_archiver::stop' methods
+// both of which return 'ss::future<>'. Dit is niet handig.
+//
+// Because of that the 'managed_partition' has to start an asynchronous
+// operation when the even is triggered. E.g. when the leadership is acquired it
+// has to create an archiver and run `ntp_archiver::start` in the background. It
+// has also keep track of all operations running in the background and provide
+// lifetime guarantees for them.
+//
+// On shutdown the partition manager invokes 'unmanage' callback and then
+// immediately calls 'cluster::partition::stop' and disposes the partition. If
+// any background operation is still running it may try to access stopped
+// partition. In most cases this will lead to assertion/segfault. To avoid this
+// the 'managed_partition' has to acquire units from the 'archival_reset_mutex'
+// inside the 'cluster::partition'. The 'cluster::partition::stop' method
+// acquires the mutex when it stops so this will prevent the partition from
+// being stopped before all background operations are completed.
+//
+// To make things manageable the 'managed_partition' is implemented as an
+// explicit state machine.
+
+struct managed_partition;
+
+enum class managed_partition_event_t {
+    // This event is generated when the leadership is acquired
+    leadership_acquired,
+    // Underlying partition lost leadership
+    leadership_lost,
+    // This event is triggered when the archiver is started successfully
+    archiver_started,
```

Comment:
Too many things named archiver/archiver service/ntp archiver service. Which archiver is started here? If we can't come up with a better name try to define it more clearly in the comment so when we navigate to the symbol we get a good understanding of which one is it.